code/xcvm/lib/core/src/gateway.rs

```diff
@@ -49,17 +59,47 @@ pub enum ExecuteMsg {
 	Bridge(BridgeMsg),
 
 	/// Message sent by an admin to register a new asset.
-	RegisterAsset {
-		asset_id: AssetId,
-		reference: AssetReference,
-	},
+	RegisterAsset(RegisterAssetMsg),
 
 	/// Message sent by an admin to remove an asset from registry.
 	UnregisterAsset {
 		asset_id: AssetId,
 	},
+
+	Wasm(Ics20MessageHook),
+}
+
+/// when message is sent to other side, we should identify receiver of some kind
+#[derive(Serialize, Deserialize, JsonSchema, Clone, PartialEq, Debug)]
+pub enum GatewayId {
+	CosmWasm(Addr),
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+
+pub struct Asset {
+	pub network_id: NetworkId,
+	pub local: AssetReference,
+	pub bridged: Option<BridgeAsset>,
+}
+
+impl Asset {
+	pub fn denom(&self) -> String {
+		self.local.denom()
+	}
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
```

Comment:
```suggestion
#[cw_serde]
#[derive(Debug)]
```

Reply:
i did not actually forked cw serde for no std, so will not work. so may other reason to upgrade forks and fix macro later